### DISCUSSION TITLE
ISO 639-2 support

### DISCUSSION THREAD

**Discussion by mposs00** [bluesky-social/atproto#1341] [Protocol (atproto)]
Notably on the Ukranian-language section of the website, there have been calls to add support for the Crimean Tatar language to bluesky's clients, such as in [this post](https://bsky.app/profile/wolfigelkott.crimea.ua/post/3k2hvhmhsk32r). Crimean Tatar, like many other indigenous languages, is under existential threat of extinction. Furthermore, there are numerous languages that are (almost arbitrarily) excluded from cataloging and tagging due to their exclusion from iso639-1. Encouraging a public discourse in these languages is not only good to help preserve the languages of these marginalized cultures, but would also allow atproto to serve a *truly* global community, regardless of whether or not their language meets the criteria for inclusion in iso639-1.

**Describe the solution you'd like**

I'd like to propose the addition of ISO 639-2 (or -3) support to the ATProto lexicons, as well as support for these languages in the bluesky client and language selector.

**Additional context**

Obviously, lack of comprehensive browser and library support for ISO 639-2 may be an obstacle. I think it is worth starting a discussion about how this deficiency can be approached, and finding the best solutions to an area where the existing implementations fail to serve significant portions of the global community.

**Comment by devinivy** [comment]
The lexicon `language` string format supports well-formed BCP 47 language tags, which should encompass the expressiveness of ISO 639-2 just fine, including the tag `crh` for the Crimean Tatar language.  You can find the validation logic around here: https://github.com/bluesky-social/atproto/blob/b9ca76f01221532e80fa71550bddf8d99e31ca12/packages/common-web/src/strings.ts#L54-L56

This doesn't cover the support you're asking for in the client, but hopefully this is already addressed down at the protocol level 👍

**Reply by mposs00** [reply]
Gotcha. I had a brief convo in discord that seemed to suggest that the support was not well-replicated across the entire atproto typescript codebase, so I wanted to follow up, but I don't have a whole lot of supporting data and haven't dug too deep into the technicalities myself, to be completely frank. Do you think this discussion would be better suited to take place on the client repo?

**Reply by mposs00** [reply]
I also suppose that I should correct a bit of a mistake in the original post, and clarify a bit; it is good to see that support is encoded into the lexicons, but does this repo implement the functionality needed for iso 639-2 handling? Or is it an issue confined solely to the client layer?

**Reply by bnewbold** [reply]
As a side note, we (I) haven't gotten around to updating our Lexicon specification docs about this recent addition (`language` string format type), which will eventually clarify how all this is expected to work.

Tracking issue: https://github.com/bluesky-social/atproto-website/issues/128

**Reply by mposs00** [reply]
Ah, thank you, this is probably why I failed to notice the specification of BCP47 for language string formats in the first place. I'm a lot more familiar with the online docs than the source of this repo.
